src/uct/ib/dc/dc_mlx5_ep.h

```diff
@@ -328,10 +395,42 @@ uct_dc_mlx5_ep_basic_init(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
                           UCS_STATS_ARG(ep->super.stats));
 }
 
-static UCS_F_ALWAYS_INLINE int
+static int
 uct_dc_mlx5_iface_dci_can_alloc(uct_dc_mlx5_iface_t *iface, uint8_t pool_index)
 {
-    return iface->tx.dci_pool[pool_index].stack_top < iface->tx.ndci;
+    uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];
+
+    if (ucs_likely(pool->stack_top < ucs_array_length(&pool->stack))) {
+        return 1;
+    }
+
+    if ((ucs_array_length(&pool->stack) == iface->tx.ndci) ||
+        (pool->stack_top == iface->tx.ndci)) {
+        return 0;
+    }
+
+    return 1;
+}
+
+static UCS_F_ALWAYS_INLINE int
+uct_dc_mlx5_iface_dci_can_alloc_or_create(uct_dc_mlx5_iface_t *iface,
+                                          uint8_t pool_index)
+{
+    int can_alloc = uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index);
+    uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];
+    ucs_status_t status;
+
+    if (!can_alloc ||
+        ucs_likely(pool->stack_top < ucs_array_length(&pool->stack))) {
```

Comment:
we already check stack_top in uct_dc_mlx5_iface_dci_can_alloc, why need again?

Reply:
We need to distinguish between 3 cases:

1. We can alloc because there are free slots nothing is need to be done
2. We can alloc by appending to the stack / dcis array
3. We can't alloc at all